### ISSUE TITLE
beetbrainz[8194]: 2022/10/02 08:15:25 Failed to decode the Emby request: unexpected end of JSON input

### ISSUE THREAD

**Issue by gbcox** [gbcox/plex-beetbrainz#2]
Trying to adapt beetbrainz to allow use with emby first and then jellyfin.  I copied plex.go and created emby.go; then changed the code to map the correct json:field names (since they are different) - but for some reason it fails with the above error message.  I've attached the webhook output from plex,  emby and this repo master branch has the updated code.   I thought this would be fairly straight forward... apparently not.  LOL.   @lyarenei I know you haven't worked on this for awhile, but thought you might know what I'm missing.  I like the idea of adding the beets info to the scrobble, so hope I can get this working for emby and then for jellyfin.  Thanks for any help you may be able to provide.

[emby_json.txt](https://github.com/gbcox/plex-beetbrainz/files/9692867/emby_json.txt)
[plex_json.txt](https://github.com/gbcox/plex-beetbrainz/files/9692869/plex_json.txt)

**Comment by lyarenei** [comment]
Going by the message alone - from general point of view, this is most certainly an issue with getting the data from the request, more specifically giving json decoder an invalid json data. So try to look into how the request has been received and what data have been parsed from the request. In general, either the json data are completely missing, are somewhere else, or you need to perform additional steps to get the data you want (this is dependent on the language/framework used).

---
Now, in this specific case, I'm pretty sure, you just need to read the data from the request body instead of form as Plex developers for some reason decided to use multipart form for their webhook data and multipart forms are usually used for uploading large amount of data and not for simple json data.

So, replacing
https://github.com/gbcox/plex-beetbrainz/blob/41567cea2faf8c0843ef1a7e0fdc75c37e339ec3/emby/emby.go#L58-L65

with (ofc with correct variable type)
https://github.com/gbcox/plex-beetbrainz/blob/41567cea2faf8c0843ef1a7e0fdc75c37e339ec3/beets/beets.go#L101-L102

should make it work.

**Comment by gbcox** [comment]
@lyarenei Thanks for the reply.  I double checked and found emby is also using multipart/form, I've included the output from webhook.site.  I'll keep looking, if you can think of anything else, please advise.  Thanks for taking the time to help.


content-length | 3326
-- | --
content-type | multipart/form-data; boundary="9b8ad52c-5073-4a22-8f86-cf042fe23ebc"
traceparent | 00-f19296cc1c0c4f407a44ea7c5b9e4645-302d56fcc718bef7-00
accept-encoding | deflate
user-agent | Emby Server/4.8.0.10
host | webhook.site

**Comment by gbcox** [comment]
Think I may have found it from looking closer at the webhook.site information.  The form value in Plex is "payload", while in Emby it's "data".   Testing and will post results here.

**Comment by gbcox** [comment]
@lyarenei that was it.  Working fine now.  Thanks again for taking the time to help.  Much appreciated!

**Comment by lyarenei** [comment]
Hmm, looks like multipart form is used much more than I thought. Anyway good to see you have figured that out!

Just wondering, if you are working with emby/jellyfin, why not use the plugin system instead of a standalone application (this)?

**Comment by gbcox** [comment]
Yeah, I'm switching back and forth between plex/emby/jellyfin.  Each has their pros and cons for me - and unfortunately they all have issues with how they perform (or not) on Roku/Google TV.  Emby, I reported their scrobbler was broken (it would dump when it encountered a track with multiple artists).  It has been over 6 months and they apparently have better things to do than to fix it.  I guess they are waiting on Listenbrainz to document how to do it, but I have given up on them ever resolving it.  Plex, of course, as you know has no plugin for Listenbrainz.  Jellyfin, you take care of the plugin, so in the case off Jellyfin, this program is not really needed - but I figure since it works with Plex and Emby now, why not add the support Jellyfin.  If people are like me and switching back and forth between the servers until they can find one they like, it's nice to have the same consistent, reliable method for scrobbling and the overhead is negligible, so why not.  Plus it's nice to pull the information from the local beets database.  My issue with Jellyfin right now is the Roku client, they just recently added support to play music with it, and it still needs major improvements.

On another topic, I haven't encountered the potential issue you mentioned regarding the missing of scrobbles.  Maybe I've just been lucky, but so far the program has been rock solid.

**Comment by lyarenei** [comment]
>  I haven't encountered the potential issue you mentioned regarding the missing of scrobbles. Maybe I've just been lucky, but so far the program has been rock solid.

Maybe something was fixed, it's been a year after all. The issue is/was that Plex didn't send the webhook request at all, not that this app would drop it. It wasn't quite frequent either, I think I found it out only by chance.
